storage/tokudb/ha_tokudb.cc

```diff
@@ -360,20 +360,40 @@ void TOKUDB_SHARE::set_cardinality_counts_in_table(TABLE* table) {
                 (i == table->s->primary_key) ||
                 (table->key_info[i].flags & HA_NOSAME);
 
-            uint32_t num_key_parts = get_key_parts(&table->key_info[i]);
-            for (uint32_t j = 0; j < num_key_parts; j++) {
+            KEY* key = &table->key_info[i];
+
+            /* Check if this index supports index statistics. */
+            if (!key->supports_records_per_key()) {
+                continue;
+            }
+
+            for (uint32_t j = 0; j < key->actual_key_parts; j++) {
+                if (j >= key->user_defined_key_parts) {
+                    key->set_records_per_key(j, 1.0);
+                    key->rec_per_key[j] = 1;
+                    continue;
+                }
+
                 assert_always(next_key_part < _rec_per_keys);
+                if ((key->flags & HA_FULLTEXT) || (key->flags & HA_SPATIAL)) {
+                    /* The record per key does not apply to
+                    FTS or Spatial indexes. */
+                    key->rec_per_key[j] = 1;
+                    key->set_records_per_key(j, 1.0);
+                    continue;
+                }
                 ulong val = _rec_per_key[next_key_part++];
-                if (is_unique_key && j == num_key_parts-1) {
+                val = (val*tokudb::sysvars::cardinality_scale_percent)/100;
+                if (val == 0)
+                    val = 1;
+
+                if (_rows == 0 ||
+                   (is_unique_key && j == key->actual_key_parts-1)) {
                     val = 1;
-                } else {
-                    val =
-                        (val*tokudb::sysvars::cardinality_scale_percent)/100;
-                    if (val == 0)
-                        val = 1;
                 }
 
-                table->key_info[i].rec_per_key[j] = val;
+                key->set_records_per_key(j, static_cast<rec_per_key_t>(val));
```

Comment:
We are computing val as integer but core server takes float, so I guess we are losing precision? Can "val" be of type rec_per_key_t (and intermediate computations reviewed for precision loss)?

Not necessarily in the scope of the current fix